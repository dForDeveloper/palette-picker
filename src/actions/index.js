export const setColors = (colors) => ({
  type: 'SET_COLORS',
  colors
});

export const setLockedColors = (lockedColors) => ({
  type: 'SET_LOCKED_COLORS',
  lockedColors
});

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const toggleLoading = (bool) => ({
  type: 'TOGGLE_LOADING',
  bool
});

export const setError = (message) => ({
  type: 'SET_ERROR',
  message
});

export const setPalettes = (id, palettes) => ({
  type: 'SET_PALETTES',
  id,
  palettes
});

export const setModal = (isDisplayed, modalType = null, text = null) => ({
  type: 'SET_MODAL',
  isDisplayed,
  modalType,
  text
});