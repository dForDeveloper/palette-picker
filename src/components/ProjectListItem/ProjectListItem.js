import React, { Component } from 'react';
import folderOpen from '../../icons/folderopen.svg';
import folderClosed from '../../icons/folderclosed.svg';
import PropTypes from 'prop-types';


export class ProjectListItem extends Component {
  renderPalettes = () => {
    return this.props.palettes.map(palette => {
      const { color1, color2, color3, color4, color5 } = palette;
      const colors = [color1, color2, color3, color4, color5];
      const paletteRectangle = colors.map(color => {
        return (
          <div
            style={{ backgroundColor: color }}
            className="Menu--palette-square"
            key={color}
          ></div>
        );
      });
      return (
        <li key={palette.created_at}>
          <p>{palette.name}</p>
          <span className="Menu--palette-span">{paletteRectangle}</span>
        </li>
      );
    })
  };

  toggleOpen = async (event) => {
    const { id } = event.target.closest('li').dataset;
    if (!this.props.palettes) await this.props.getPalettes(id);
    const { name, opened } = this.props;
    this.props.toggleOpen(name, !opened);
  };

  render () {
    const { id, name, opened } = this.props;
    return (
      <li data-id={id} className="Menu--list-item">
        {!opened && 
          <img
            src={folderClosed}
            alt="closed folder"
            className="Menu--icon"
            onClick={this.toggleOpen} 
          />
        }
        {opened && 
          <img
            src={folderOpen}
            alt="open folder"
            className="Menu--icon"
            onClick={this.toggleOpen} 
          />
        }
        <span onClick={this.toggleOpen}>{name}</span>
        {opened && 
          <ul className="Menu--ul-palettes">{this.renderPalettes()}</ul>}
      </li>
    );
  }
}

ProjectListItem.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  getPalettes: PropTypes.func
};