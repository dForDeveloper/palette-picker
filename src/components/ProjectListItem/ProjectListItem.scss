.Menu--list-item {
  cursor: pointer;
  margin-bottom: 8px;
}

.Menu--icon {
  vertical-align: middle;
  margin-right: 8px;
}

.Menu--ul-palettes {
  list-style-type: none;
  margin-left: 40px;
  margin-right: 16px;
}

.Menu--palette-span {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  border: 1px solid #ffffff;
  margin-top: 2px;
}

.Menu--palette-square {
  width: 100%;
  height: 16px;
  display: inline-block;
}