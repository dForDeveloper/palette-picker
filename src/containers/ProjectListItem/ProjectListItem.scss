.Menu--list-item {
  margin-bottom: 8px;
  background-color: #ffffff;
  color: #000000;
  transition: all 0.2s ease;
  &:hover {
    background-color: #0facf3;
    color: #ffffff;
  }
}

.Menu--icon-folder {
  vertical-align: middle;
  margin-right: 8px;
  cursor: pointer;
}

.Menu--icon-edit {
  vertical-align: middle;
  cursor: pointer;
  height: 20px;
  float: right;
  margin-top: 2px;
}

.Menu--icon-edit-light {
  @extend .Menu--icon-edit;
  height: 18px;
  margin-top: 0;
}

.Menu--project-name {
  @extend .Menu--text;
  height: 19px;
  width: 168px;
  cursor: pointer;
}