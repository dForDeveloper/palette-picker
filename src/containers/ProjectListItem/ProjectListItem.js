import React, { Component } from 'react';
import { setModal } from '../../actions';
import folder from '../../icons/folder.svg';
import editdark from '../../icons/editdark.svg';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

export class ProjectListItem extends Component {
  toggleOpen = async (event) => {
    const { id } = event.target.closest('li').dataset;
    if (!this.props.palettes) await this.props.getPalettes(id);
    this.props.setActiveProject(id)
    this.props.toggleDropDown();
  };

  setModal = () => {
    this.props.setModal(true, 'project', this.props.name);
  }

  render () {
    const { id, name } = this.props;
    return (
      <li data-id={id} className="Menu--list-item">
        <img
          src={folder}
          alt="folder icon"
          className="Menu--icon-folder"
          onClick={this.toggleOpen} 
        />
        <p className="Menu--project-name" onClick={this.toggleOpen}>
          {name}
        </p>
        <img
          src={editdark}
          alt="edit icon"
          className="Menu--icon-edit"
          onClick={this.setModal}
        />
      </li>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  setModal: (isDisplayed, type, name) => {
    dispatch(setModal(isDisplayed, type, name))
  }
});

export default connect(null, mapDispatchToProps)(ProjectListItem);

ProjectListItem.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  getPalettes: PropTypes.func,
  setModal: PropTypes.func
};