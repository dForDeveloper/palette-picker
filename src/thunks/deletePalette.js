import { setError, toggleLoading, setModal, removePalette } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const deletePalette = (projectID, paletteID) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const options = createOptions('DELETE');
      await fetchData(`/api/v1/palettes/${paletteID}`, options);
      dispatch(removePalette(projectID, paletteID));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error deleting palette: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}