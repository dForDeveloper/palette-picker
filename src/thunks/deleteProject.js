import { setError, toggleLoading, removeProject, setModal } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const deleteProject = (id) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const options = createOptions('DELETE');  
      await fetchData(`/api/v1/projects/${id}`, options);
      dispatch(removeProject(id));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error deleting project: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}