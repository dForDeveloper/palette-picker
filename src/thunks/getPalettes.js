import { fetchData } from '../utils/api';
import { toggleLoading, setError, setPalettes } from '../actions';

export const getPalettes = (id) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const palettes = await fetchData(`/api/v1/projects/${id}/palettes`);
      dispatch(setPalettes(id, palettes));
    } catch (error) {
      dispatch(setError('Error getting palettes: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
};