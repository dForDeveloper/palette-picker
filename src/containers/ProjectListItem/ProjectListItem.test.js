import React from 'react';
import { shallow } from 'enzyme';
import { setModal } from '../../actions';
import { getPalettes } from '../../thunks/getPalettes';
import { ProjectListItem, mapDispatchToProps } from './ProjectListItem';

jest.mock('../../thunks/getPalettes');

describe('ProjectListItem', () => {
  let wrapper;
  const mockProps = {
    id: 1,
    name: 'my project',
    getPalettes: jest.fn(),
    setModal: jest.fn(),
    setActiveProject: jest.fn(),
    toggleDropDown: jest.fn()
  };

  beforeEach(() => {
    wrapper = shallow(<ProjectListItem {...mockProps} />);
  });

  describe('render', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call setModal when the edit icon is clicked', () => {
      jest.spyOn(wrapper.instance(), 'setModal');
      wrapper.instance().forceUpdate();
      wrapper.find('.Menu--icon-edit').simulate('click');
      expect(wrapper.instance().setModal).toHaveBeenCalled();
    });
    
    it('should call toggleOpen when the folder icon is clicked', () => {
      jest.spyOn(wrapper.instance(), 'toggleOpen');
      wrapper.instance().forceUpdate();
      wrapper.find('.Menu--icon-folder').simulate('click');
      expect(wrapper.instance().toggleOpen).toHaveBeenCalled();
    });
    
    it('should call toggleOpen when the name is clicked', () => {
      jest.spyOn(wrapper.instance(), 'toggleOpen');
      wrapper.instance().forceUpdate();
      wrapper.find('.Menu--project-name').simulate('click');
      expect(wrapper.instance().toggleOpen).toHaveBeenCalled();
    });
  });

  describe('toggleOpen', () => {
    it('should call three functions', async () => {
      await wrapper.instance().toggleOpen();
      expect(mockProps.getPalettes).toHaveBeenCalledWith(1);
      expect(mockProps.setActiveProject).toHaveBeenCalledWith(1);
      expect(mockProps.toggleDropDown).toHaveBeenCalled();
    });
  });

  describe('setModal', () => {
    it('should call setModal', () => {
      wrapper.instance().setModal();
      expect(mockProps.setModal)
      .toHaveBeenCalledWith(true, 'project', 'my project', 1);
    });
  });
});

describe('mapDispatchToProps', () => {
  const mockDispatch = jest.fn();
  const mappedProps = mapDispatchToProps(mockDispatch);

  it('should call getPalettes', () => {
    const expected = getPalettes(1);
    mappedProps.getPalettes(1);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
  
  it('should call setModal', () => {
    const expected = setModal(true, 'project', 'my project', 1);
    mappedProps.setModal(true, 'project', 'my project', 1);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});