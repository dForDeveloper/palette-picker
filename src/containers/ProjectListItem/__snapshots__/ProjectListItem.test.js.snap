// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectListItem render should match the snapshot 1`] = `
<li
  className="Menu--list-item"
>
  <img
    alt="folder icon"
    className="Menu--icon-folder"
    onClick={[Function]}
    src="folder.svg"
  />
  <p
    className="Menu--project-name"
    onClick={[Function]}
  >
    my project
  </p>
  <img
    alt="edit icon"
    className="Menu--icon-edit"
    onClick={[Function]}
    src="editdark.svg"
  />
</li>
`;
